webapp/app/reducers/usersReducer.js

```diff
@@ -1,90 +1,79 @@
-import { getUserId } from '../utils/user';
-
-export const initialState = {
-  data: [],
-  selectedUser: {},
-  fetch: {
-    loading: false,
-    error: null
-  }
-};
-
-export default function usersReducer(state = initialState, action) {
-  switch(action.type) {
-    case CREATE_USERS_SUCCESS:
-      return {
-        ...state,
-        data: [
-          action.payload,
-          ...state.data
-        ]
-      };
-
-    case DELETE_USERS_SUCCESS:
-      return {
-        ...state,
-        data: state.data.filter(user => getUserId(action.payload) !== getUserId(user))
-      };
-
-    case GET_USERS_SUCCESS:
-      return {
-        ...state,
-        data: action.payload.users
-      };
-
-    case SELECT_USERS_SUCCESS:
-      return {
-        ...state,
-        selectedUser: action.payload
-      };
-
-    case UPDATE_USERS_SUCCESS:
-      return {
-        ...state,
-        data: [
-          action.payload,
-          ...state.data.filter(user => getUserId(action.payload) !== getUserId(user))
-        ]
-      };
-
-    case LOADING_USERS_BEGIN:
-      return {
-        ...state,
-        fetch: {
-          loading: true,
-          error: null
-        }
-      };
-
-    case LOADING_USERS_COMPLETE:
-      return {
-        ...state,
-        fetch: {
-          loading: false,
-          error: null
-        }
-      };
-
-    case LOADING_USERS_FAILED:
-      return {
-        ...state,
-        fetch: {
-          loading: false,
-          error: action.payload.error
-        }
-      };
+import { getUserId } from '../utils';
+import { createReducer } from '../utils';
+import initialState from './initialState';
 
-    default:
-      return state;
+export const users = createReducer(initialState.users, {
+  [GET_USERS_SUCCESS](state, { users }) {
+    return {
+      ...state,
+      data: users
+    };
+  },
+  [CREATE_USERS_SUCCESS](state, { user }) {
+    return {
+      ...state,
+      data: [
+        user,
+        ...state.data
+      ]
+    };
+  },
+  [DELETE_USERS_SUCCESS](state, { user }) {
+    return {
+      ...state,
+      data: state.data.filter(sourceUser => getUserId(user) !== getUserId(sourceUser))
+    };
+  },
+  [UPDATE_USERS_SUCCESS](state, { user }) {
+    return {
+      ...state,
+      data: [
+        user,
+        ...state.data.filter(sourceUser => getUserId(user) !== getUserId(sourceUser))
+      ]
+    };
+  },
+  [SELECT_USERS_SUCCESS](state, { user }) {
+    return {
+      ...state,
+      selectedUser: user
+    };
+  },
+  [LOADING_USERS_BEGIN](state) {
+    return {
+      ...state,
+      fetch: {
+        loading: true,
+        error: null
+      }
+    };
+  },
+  [LOADING_USERS_COMPLETE](state) {
+    return {
+      ...state,
+      fetch: {
+        loading: false,
+        error: null
+      }
+    };
+  },
+  [LOADING_USERS_FAILED](state, { error }) {
+    return {
+      ...state,
+      fetch: {
+        loading: false,
+        error: error
+      }
+    };
   }
-}
+});
```

Comment:
This ended up very similar to the handleActions approach:

```
export default handleActions({
  [`${USERS.GET_ALL}_PENDING`]: state => ({
    ...state,
    fetch: {
      ...state.fetch,
      pending: true
    }
  }),
  [`${USERS.GET_ALL}_FULFILLED`]: (state, action) => ({
    ...state,
    data: action.payload.data.docs,
    fetch: usersInitialState.fetch
  }),
  [`${USERS.GET_ALL}_REJECTED`]: state => ({
    ...state,
    fetch: {
      ...usersInitialState.fetch,
      error: true
    }
  }),
  [USERS.SELECT]: (state, action) => {
    return ({
      ...state,
      selectedUser: action.payload
    });
  },
  [`${USERS.CREATE}_PENDING`]: state => ({
    ...state,
    fetch: {
      ...state.fetch,
      pending: true
    }
  }),
  [`${USERS.CREATE}_FULFILLED`]: (state, action) => ({
    ...state,
    data: [
      action.payload.data,
      ...state.data
    ],
    fetch: usersInitialState.fetch
  }),
  [`${USERS.CREATE}_REJECTED`]: state => ({
    ...state,
    fetch: {
      ...usersInitialState.fetch,
      error: true
    }
  }),
  [`${USERS.UPDATE}_PENDING`]: state => ({
    ...state,
    fetch: {
      ...state.fetch,
      pending: true
    }
  }),
  [`${USERS.UPDATE}_FULFILLED`]: (state, action) => {
    const targetUser = action.payload.data;
    
    return ({
      ...state,
      data: [
        targetUser,
        ...state.data.filter(user => getUserId(targetUser) !== getUserId(user))
      ],
      fetch: usersInitialState.fetch
    });
  },
  [`${USERS.UPDATE}_REJECTED`]: state => ({
    ...state,
    fetch: {
      ...usersInitialState.fetch,
      error: true
    }
  }),
  [`${USERS.DELETE}_PENDING`]: state => ({
    ...state,
    fetch: {
      ...state.fetch,
      pending: true
    }
  }),
  [`${USERS.DELETE}_FULFILLED`]: (state, action) => ({
    ...state,
    data: state.data.filter(user => getUserId(action.payload) !== getUserId(user)),
    fetch: usersInitialState.fetch
  }),
  [`${USERS.DELETE}_REJECTED`]: state => ({
    ...state,
    fetch: {
      ...usersInitialState.fetch,
      error: true
    }
  })
}, usersInitialState);
```

Reply:
This is the sample implementation of redux-saga from their provided guides, please check their guides and docs, for insight into why this is implemented in this way, thanks

Reply:
what is the actionable item here? Please provide actionable clear solutions or changes to perform or how would you implement it and Why, thanks

Reply:
What I meant is that this code ended up similar to the boilerplate code that I did and was removed.

Reply:
ah gotcha, it is somewhat similar but not the same though, we are not using promise middleware nor redux's createAction/handleAction coupled with promise middleware, the rest of the implementation is redux standard along with es6 syntax, but point taken looks similar.

are you just pointing it out, or would you prefer we change it? and if so what would be the driver to change it? thanks